api/src/services/roles.ts

```diff
@@ -181,49 +187,200 @@ export class RolesService extends ItemsService {
 	override async createOne(data: Partial<Item>, opts?: MutationOptions): Promise<PrimaryKey> {
 		this.assertValidIpAccess(data);
 
+		const increasedCounts: AccessTypeCount = {
+			admin: 0,
+			app: 0,
+			api: 0,
+		};
+
+		if ('users' in data) {
+			if ('admin_access' in data && data['admin_access'] === true) {
+				increasedCounts.admin += data['users'].length;
+			} else if ('app_access' in data && data['app_access'] === true) {
+				increasedCounts.app += data['users'].length;
+			} else {
+				increasedCounts.api += data['users'].length;
+			}
+		}
+
+		await checkIncreasedUserLimits(this.knex, increasedCounts);
+
 		return super.createOne(data, opts);
 	}
 
 	override async createMany(data: Partial<Item>[], opts?: MutationOptions): Promise<PrimaryKey[]> {
+		const increasedCounts: AccessTypeCount = {
+			admin: 0,
+			app: 0,
+			api: 0,
+		};
+
 		for (const partialItem of data) {
 			this.assertValidIpAccess(partialItem);
+
+			if ('users' in partialItem) {
+				if ('admin_access' in partialItem && partialItem['admin_access'] === true) {
+					increasedCounts.admin += partialItem['users'].length;
+				} else if ('app_access' in partialItem && partialItem['app_access'] === true) {
+					increasedCounts.app += partialItem['users'].length;
+				} else {
+					increasedCounts.api += partialItem['users'].length;
+				}
+			}
 		}
 
+		await checkIncreasedUserLimits(this.knex, increasedCounts);
+
 		return super.createMany(data, opts);
 	}
 
 	override async updateOne(key: PrimaryKey, data: Partial<Item>, opts?: MutationOptions): Promise<PrimaryKey> {
 		this.assertValidIpAccess(data);
 
 		try {
+			const increasedCounts: AccessTypeCount = {
+				admin: 0,
+				app: 0,
+				api: 0,
+			};
+
+			let increasedUsers = 0;
+
+			const existingRole:
+				| { count: number | string; admin_access: number | boolean; app_access: number | boolean }
+				| undefined = await this.knex
+				.count('directus_users.id', { as: 'count' })
+				.select('directus_roles.admin_access', 'directus_roles.app_access')
+				.from('directus_users')
+				.where('directus_roles.id', '=', key)
+				.andWhere('directus_users.status', '=', 'active')
+				.leftJoin('directus_roles', 'directus_users.role', '=', 'directus_roles.id')
+				.first();
```

Comment:
This query currently returns `{ count: 0, admin_access: null, app_access: null }` for non-existent roles, not `undefined`.

Reply:
Hmmm that renders this check obsolete https://github.com/directus/directus/pull/22479#discussion_r1624641271

Reply:
Yeah, it would probably just work like this, but we should at least update the expected types.

Reply:
Okay, this was for SQLite, on PostgreSQL it doesn't work at all (even for existing roles) 🚨 

```
select count("directus_users"."id") as "count", "directus_roles"."admin_access", "directus_roles"."app_access" from "directus_users" left join "directus_roles" on "directus_users"."role" = "directus_roles"."id" where "directus_roles"."id" = $1 and "directus_users"."status" = $2 limit $3 - column "directus_roles.admin_access" must appear in the GROUP BY clause or be used in an aggregate function
```

Reply:
with the `groupBy` it does indeed return `undefined`